### ISSUE TITLE
Speech announcement improvement

### ISSUE THREAD

**Issue by PinkDuck** [osmandapp/Osmand#1411]
While driving, I sometimes hear "Prepare to turn left in 3 km at the roundabout", which causes amusement as it's a pretty unhelpful outburst. A better voice instruction might be "Continue for 3 km to roundabout" (always put distances first in vocal statements).

**Comment by sonora** [comment]
I agree to some degree. But if we wanted to change this, we need a full concept replacing all "Prepare to..." prompts by better wording. So what do we do use for the current prompts like ""Prepare to turn left in 3km"?

It would be helpful to do a little research of what convention other nav systems use here.

**Comment by breversa** [comment]
Such as "In 3km, prepare to turn left at the roundabout" ?

**Comment by PinkDuck** [comment]
At 3 km distance away the idea of 'preparing' for a turn at the roundabout seems laughable, as often there's absolutely nothing to do if on a single lane. The information that there's a roundabout ahead in 3 km is however useful for general awareness. Strangely nothing is said about mini-roundabouts.

**Comment by PinkDuck** [comment]
I used HERE navigation for Android KitKat today and what a relief it was in terms of voice guidance compared to OsmAnd. No hurried speech, no cluttered together instructions, just clear well designed guidance. From what I recall, these are the statements issued:

After 600 m, turn right
Now turn right
After 200 m, at the roundabout, take the 2nd exit
At the roundabout, take the 2nd exit
Now take the 2nd exit
Follow the course of the road for 9 km
After 1 km turn left
After 300 m, turn left then immediately right
Turn right then turn left
After 200 m, keep left
Now keep left

Note how this follows the distance first principle, except for when there's no instructions for a lengthy period. It also mentions nothing about the action to take place after the lengthy distance.

**Comment by sonora** [comment]
I agree we still need to tweak the concept of what prompts we need (and perhaps also when they should sound). I had started this a long time ago, but got distracted by other things. I had documented the status where I left it at here: http://code.google.com/p/osmand/wiki/VoiceRoutingCurrentStatus
(Some recent updates may never have made it into this old documentation.)

Please understand how our triggers are a mix of distance and lead time, depending what comes first.

If anybody could update the tables as a suggestion and post here, we can review together, then I can implement the improvement very quickly!

Thanks, Hardy

**Comment by sonora** [comment]
I have updated the documentation, see attachment, and made some changes to the effect that the PREPARE_LONG (3km for cars) voice prompts are not played now, we can test in the nightlies.

It is not perfect yet, I will have to go through it again (looks like we still have the "Follow the course of the road..." prompt too much), I will continue to debug and fine-tune in the next few days.

Best, Hardy
![voiceturnprompts_2015-09-04](https://cloud.githubusercontent.com/assets/845735/9695281/4c7c480e-5360-11e5-8323-a6ae7f1bd9da.png)

**Comment by PinkDuck** [comment]
I would like to see the prefix always simplified to just "In x meters" instead of "In / After x meters", since that's one fewer syllable to comprehend. Or the "After" switched to the further out instruction.

**Comment by sonora** [comment]
I have switched "After" to the further out prompt, and also slightly increased the minimum spacing between the MAKE_TURN_IN and MAKE_TURN instruction. Picture above is updated. Let's see now.

**Comment by PinkDuck** [comment]
Something also needs to be done to avoid drivers being overloaded with too much information. I actually developed a headache after trying to comprehend multiple instructions while driving in a tricky unknown area with many street parked cars. Things such as "Turn right then after 200 meters stay left, the trip is about 2 km, time 9 minutes, turn right then after 200 meters stay left, attention pedestrian crosswalk" when all I wanted to hear was "turn right".

Similarly I don't need to know the name of the road for the exit at a roundabout, just the exit number is sufficient. This also should be determined in advance so that GPS position error can't result in hearing the wrong number exit to take on approach!

**Comment by vshcherb** [comment]
Prepare long makes perfect sense for long running distances, when you drive > 50 km straight, these messages could attract required attention.

**Comment by aceman444** [comment]
PinkDuck, I think you can disable speaking of street names or pedestrian crossing in OsmAnd settings.

**Comment by sonora** [comment]
Ok, thanks for the feedback, let's take this one by one:

(1) In tight city driving (like in residential areas), when you deviate from the route you always hear the full "Route recalculated with new distance and new time", which is actually a bug. For the fast-paced car mode only a quick "route recalculated" should sound (making you aware you deviated from where told to go.) This is in the works, and will definitely "de-crowd" these messages, particularly in connection with (2):

(2) On some occasions, the prompts for MAKE_TURN_IN and MAKE_TURN move very close together, but I think this may mainly result from issue (1) above, so I would like to re-evaluate after fixing (1).

(3) "then after ..." in the MAKE_TURN and MAKE_TURN_IN is only announced if the second turn is very close after the first one (currently 200 m). I guess this is useful for driving and should be kept.  (?) Or should it be disabled in one of the?   (Or maybe leave out street names for the then-after prompt?)

(4) Street names: Yes, these make prompts longer and more "crowded". Their usability varies by country: They are REALLY helpful in some countries where big street name markers are posted above intersections, like in US cities, while they may be less useful in some European countries where we announce some street ref number which is not even visible on street signs. So I guess what aceman444 remarks is key here: Turn them off in the settings where you find them not useful?

(5) Street names for roundabout exits: I have less experience with these, we could disable them, but would need a common agreement tested in different countries.

(6) PREPARE_LONG: Victor is right, these are for very long haul drives to "start awareness". In the US, pre-announcements for motorway intersections are posted 2 miles out, in Europe mostly 2 km out. PREPARE_LONG made sure you receive the "wakeup" before these signs pass by.
Signs for motorway exits, on the other hand, are posted 1 mile ahead (US) and 1km ahead (Europe), so these are close to our "PREPARE" prompt.

Most other nav systems I checked have no PREPARE_LONG equivalent, and from my testing the last few days it may actually be expendable. For people who want frequent prompting, we actually have a "repeat nav announcements" setting, but I have never used or tested it.
 
(7) Another occasional issue is that we still do not seem to stop (or even cut while playing) prompts which are outdated, I am saying that our wrongMovementDirection switch may have room for improvement: You can test this by deliberately missing a motorway exit, going straight on. After hearing the "TURN" prompt to take the highway exit (the one you are ignoring), you may still here the following prompt for the turn after this, even if the highway exit made a 180 degree loop "backwards" before getting to the next intersection, while you are moving "forward" on the motorway. Only after that "obsolete prompt" you will hear  "Route recalculated". I am not sure if we queue the prompt too early and never flush it when it becomes invalid, or if there is a trickier problem with detecting that/when we are not on the route anymore.

**Comment by PinkDuck** [comment]
While aware I can disable street names and hazard announcements I do like these for identifying the street I am turning on to is correct and for heads up while driving before not always easy to spot crossings. The set of instructions from my previous post were all triggered when I was stationary in a car park beside a road though just about to start off again.

**Comment by Gizmot** [comment]
1. My biggest gripe is with your car profile table.
You should have multiple car tables depending on speed. In a city street, going 50km/h, getting some information about a turn 1.5 km away is simply bad.

My suggestion would be to have multiple tables:
0-51km/km
52-85km
85-110
110+

You could probe for street speed limit, if not available, use current GPS speed.
(You need to make sure not to repeat instruction if you are slowing down and changing speed profile :D)

2. On an highway, instead of "turn right", I much prefer "exit right".

3. It drives me mad whenever I hear "Follow the course of the road for...". What about "keep going for..."
short and sweet

4. The " the trip is about 2 km, time 9 minutes" is completely unnecessary. All it does is makes you look weird when you simulated a route in a public place.

5. What is that "warning: stop sign" clue? I get that once in a while and confusing the shit out of me.

**Comment by sonora** [comment]
Gizmot, thanks for your feedback! A few comments here, since you asked:

(A) Please note that no case distinction by speed intervals is needed, as the timing of the announcements is fully (linearly) speed adjusted already (that's what the lead time intervals mean to indicate). Only the far-out turn prompt (C) is defined by distance only, as for far-out prompts shifting them by speed makes not much difference anyway. These prompts hardly come into play in city driving, or if they do they give you some confidence that you now need to stick with your current road for an extensive period, which I reassuring to know in unfamiliar territory, I would say?

(B) True, but currently we do not have any good "exit detection" in our code (often not in the map data, either.)  Ultimate goal is be to have "Take the exit" where appropriate - but that is future-ware for now.

(C) Debatable. I personally tend to say you are right, but almost all nav systems I have come across seem to use "Follow the course ...", so for now I went with what apparently is more conventional. Maybe when we get to supporting American vs. British vs. other versions of English we can do this more easily ... :-)

(D) Not sure what you are saying here. I think that whenever a route is first calculated this is essential information people want to hear while starting to drive?

(E) You can disable these warnings in the nav settings. I think we may still have the bug that also stop signs of intersecting roads (not your 4-way stops, or 2-way stops in exactly your path) may be announced, so maybe that's what's confusing here.

**Comment by PinkDuck** [comment]
I suspect "Follow the course of the road" was chosen as it sounds lengthy and there's time to spare to say it and without even listening to the words the driver would definitely recognise it as there's nothing further for a while. However, the advice could even be shortened further to just "[Now ]continue for...".

**Comment by sonora** [comment]
Fair enough. It is in the new nightly, let's test it  :-)

Best, Hardy

**Comment by Gizmot** [comment]
(B) Could it be a good idea to also use something like "take ramp right" instead of "slightly turn right".
I often get to "slightly" turn. In my mind, "slightly" should be very rare cases of a Y fork.
(C) When I use a GPS with voice announcement, I want a rally pro co-driver. Short, efficient and surgical. Hearing "Follow the course the course of the road for six hundred and forty meters" takes about 4 seconds. That means that as soon as I hear "Follow..." I need to keep a back process in head to understand  the actual usefull information 3 seconds later. At 100km/h, I'm about 100m down the street already. Also, you could probably round down or up up to the closest 100m. Short and sweet. 
(D) That is more or less (in my opinion) voice that could be eliminated. That sounds more like noise pollution to me. You should beforehand have a pretty good idea in your head of this information. If not (completely unknown way , you should be double-checking your distance/path  and your screen will be in your face anyway.
(E) Could be that, because there is never any stop sign in place when I hear that.

**Comment by breversa** [comment]
> (B) Could it be a good idea to also use something like "take ramp right" instead of "slightly turn right".
I often get to "slightly" turn. In my mind, "slightly" should be very rare cases of a Y fork.

Same here : I'm often confused by "slightly turn right" and "turn right", and still don't know the difference between both. Is it a matter of angle ?

I'm also often wondering if "Keep right" means "stay on the right lane where I'm already" or "switch to the next right/exit lane". Some other navigation devices have "Take the [first/second/third] lane on the right" which I find much clearer.

**Comment by sonora** [comment]
Yes, "turn slightly" is a matter of angle (in analogy to turn sharply), and confused many. I have now changed the limit angle t<45 degrees (not sure why we had set it at 60, I think that as  bug.) It should really only come up at forks, and at exit ramps (for which we have no good separate prompt yet.)

Folks, let me close tis issue, I have been driving around with the changes for a few hundred miles and think we have achieved a lot! The original issue has been solved, Should you have any more specific ideas, it is likely good to open a new issue.

Brief summary on what was done:
- Silenced the extreme long warning, which annoyed many  (original issue 1411)
- Cleaned up some surplus code and eliminated minor bugs
- Consistent "distance first" now in all prompts (en at least)
- Fixed issue with tight prompts due to time and distance announcement upon route recalculation
- Better minimum spacing between TURN_IN and TURN prompt
- Shortened "Follow the course of  the road... " to "Now continue"
- Better restricted "slight turn" to forks and exits, which also shortens many prompts

I think we have reached what with reasonable effort can be done with the current engine and available turn types, but certainly I will keep my eyes open for further possible fine tuning on my trips  :-)

**Comment by Gizmot** [comment]
"with reasonable effort can be done with the current engine and available turn types"
Is there any documentation anywhere that describe the capabilities and inner working of the current engine? For exemple, while on a street that is not an highway=motorway, turning to any   highway=motorway_link could be a ramp. Is that currently possible?

**Comment by sonora** [comment]
You'd have to look at the code, I'm afraid, to see what types of turns and other announcements we detect and send to the voice prompter. No good "Exit" detection so far, as I had already mentioned above.
